types/src/error.rs

```diff
@@ -177,32 +177,47 @@ impl Error {
 	}
 }
 
-/// Error type with a special `subscription_closed` field to detect that
+/// A type with a special `subscription_closed` field to detect that
 /// a subscription has been closed to distinguish valid items produced
 /// by the server on the subscription stream from an error.
+///
+/// This is included in the `result field` of the SubscriptionResponse
+/// when an error is reported by the server.
 #[derive(Deserialize, Serialize, Debug, PartialEq)]
-pub struct SubscriptionClosedError {
-	subscription_closed: String,
-	id: u64,
+pub struct SubscriptionClosed {
+	reason: SubscriptionClosedReason,
 }
 
-impl SubscriptionClosedError {
-	/// Create a new subscription closed error.
-	pub fn new(reason: impl Into<String>, id: u64) -> Self {
-		Self { subscription_closed: reason.into(), id }
+impl From<SubscriptionClosedReason> for SubscriptionClosed {
+	fn from(reason: SubscriptionClosedReason) -> Self {
+		Self::new(reason)
 	}
+}
 
-	/// Get the reason why the subscription was closed.
-	pub fn close_reason(&self) -> &str {
-		&self.subscription_closed
+impl SubscriptionClosed {
+	/// Create a new [`SubscriptionClosed`].
+	pub fn new(reason: SubscriptionClosedReason) -> Self {
+		Self { reason }
 	}
 
-	/// Get the subscription ID.
-	pub fn subscription_id(&self) -> u64 {
-		self.id
+	/// Get the close reason.
+	pub fn close_reason(&self) -> &SubscriptionClosedReason {
+		&self.reason
 	}
 }
 
+/// A type to represent when a subscription gets closed
+/// by either the server or client side.
+#[derive(Deserialize, Serialize, Debug, PartialEq)]
```

Comment:
//cc @maciejhirsz @dvdplm  as we already changed `close_reason` -> `reason`

should have some obscure `#[serde(tag = "_secret_tag")]` such that the type get mixed up with a something else?

Reply:
right, I think `tag doesn't` work on new type pattern nvm i.e, Server(String)